src/main/resources/js/macro-customise.js

```diff
@@ -34,4 +34,110 @@ AJS.bind('init.rte', function () {
     }
   });
 
-});
\ No newline at end of file
+});
+
+AJS.$(document).ready(function () {
+  var pageId = AJS.Meta.get("page-id");
+  var baseUrl = AJS.Confluence.getBaseUrl();
+  // Get all shadow elements from page
+  // return a dom node list
+  function getAllShadowElements() {
+    var elements = [];
+    AJS.$('sequence-diagram').each(function () {
+      if (this.shadowRoot.querySelector(".sequence-diagram-container")) {
+        elements.push(this.shadowRoot.querySelector(".sequence-diagram-container"))
+      }
+    });
+    return elements;
+  }
+  // Get all Macro nodes and its dsl
+  // return dsl text and convertToBlob promise
+  function getConvertPromiseAndDsl() {
+    var doms = getAllShadowElements();
+    var objects = [];
+    if (doms.length > 0) {
+      var dslList = [];
+      AJS.$.each(doms, function (index, dom) {
+        var dslEls = AJS.$(dom).find('.zenuml-dsl');
+        if (dslEls.length > 0) {
+          var dslText = dslEls.text();
+          // remove duplicate when upload mutiple same uml at the same time
+          if (dslList.indexOf(dslText) < 0) {
+            objects.push({
+              dsl: dslText,
+              cpro: domtoimage.toBlob(dom, { bgcolor: 'white' })
```

Comment:
what does `cpro` mean?

Reply:
means convert promise, I still cannot figure out a good name for it.